runtime/src/iree/builtins/ukernel/arch/arm_64/mmt4d_arm_64_dotprod.c

```diff
@@ -75,3 +75,129 @@ IREE_UK_MMT4D_TILE_FUNC_IMPL_FOR_M0_1_2_4_8(
+      int8x8_t lhs_1 = vget_low_s8(lhs_uzp[0].val[1]);
+      int8x8_t lhs_2 = vget_high_s8(lhs_uzp[0].val[0]);
+      int8x8_t lhs_3 = vget_high_s8(lhs_uzp[0].val[1]);
+      int8x8_t lhs_4 = vget_low_s8(lhs_uzp[1].val[0]);
+      int8x8_t lhs_5 = vget_low_s8(lhs_uzp[1].val[1]);
+      int8x8_t lhs_6 = vget_high_s8(lhs_uzp[1].val[0]);
+      int8x8_t lhs_7 = vget_high_s8(lhs_uzp[1].val[1]);
+
+      acc[0] = vdotq_lane_s32(acc[0], rhs_0, lhs_0, 0);
+      acc[1] = vdotq_lane_s32(acc[1], rhs_0, lhs_0, 1);
+      acc[2] = vdotq_lane_s32(acc[2], rhs_0, lhs_2, 0);
+      acc[3] = vdotq_lane_s32(acc[3], rhs_0, lhs_2, 1);
+      acc[4] = vdotq_lane_s32(acc[4], rhs_0, lhs_4, 0);
+      acc[5] = vdotq_lane_s32(acc[5], rhs_0, lhs_4, 1);
+      acc[6] = vdotq_lane_s32(acc[6], rhs_0, lhs_6, 0);
+      acc[7] = vdotq_lane_s32(acc[7], rhs_0, lhs_6, 1);
+
+      acc[0] = vdotq_lane_s32(acc[0], rhs_1, lhs_1, 0);
+      acc[1] = vdotq_lane_s32(acc[1], rhs_1, lhs_1, 1);
+      acc[2] = vdotq_lane_s32(acc[2], rhs_1, lhs_3, 0);
+      acc[3] = vdotq_lane_s32(acc[3], rhs_1, lhs_3, 1);
+      acc[4] = vdotq_lane_s32(acc[4], rhs_1, lhs_5, 0);
+      acc[5] = vdotq_lane_s32(acc[5], rhs_1, lhs_5, 1);
+      acc[6] = vdotq_lane_s32(acc[6], rhs_1, lhs_7, 0);
+      acc[7] = vdotq_lane_s32(acc[7], rhs_1, lhs_7, 1);
+    } else if (M0 == 4) {
+      // 4x2 * 2x4 -> 4x4.
+      int8x16x2_t lhs_uzp = vld2q_s8(lhs_ptr);
+      lhs_ptr += 32;
+
+      int8x8_t lhs_0 = vget_low_s8(lhs_uzp.val[0]);
+      int8x8_t lhs_1 = vget_low_s8(lhs_uzp.val[1]);
+      int8x8_t lhs_2 = vget_high_s8(lhs_uzp.val[0]);
+      int8x8_t lhs_3 = vget_high_s8(lhs_uzp.val[1]);
+
+      acc[0] = vdotq_lane_s32(acc[0], rhs_0, lhs_0, 0);
+      acc[1] = vdotq_lane_s32(acc[1], rhs_0, lhs_0, 1);
+      acc[2] = vdotq_lane_s32(acc[2], rhs_0, lhs_2, 0);
+      acc[3] = vdotq_lane_s32(acc[3], rhs_0, lhs_2, 1);
+
+      acc[0] = vdotq_lane_s32(acc[0], rhs_1, lhs_1, 0);
+      acc[1] = vdotq_lane_s32(acc[1], rhs_1, lhs_1, 1);
+      acc[2] = vdotq_lane_s32(acc[2], rhs_1, lhs_3, 0);
+      acc[3] = vdotq_lane_s32(acc[3], rhs_1, lhs_3, 1);
+    } else if (M0 == 2) {
+      // 2x2 * 2x4 -> 2x4.
+      int8x8x2_t lhs_uzp = vld2_s8(lhs_ptr);
+      lhs_ptr += 16;
+
+      acc[0] = vdotq_lane_s32(acc[0], rhs_0, lhs_uzp.val[0], 0);
+      acc[1] = vdotq_lane_s32(acc[1], rhs_0, lhs_uzp.val[0], 1);
+
+      acc[0] = vdotq_lane_s32(acc[0], rhs_1, lhs_uzp.val[1], 0);
+      acc[1] = vdotq_lane_s32(acc[1], rhs_1, lhs_uzp.val[1], 1);
+    } else if (M0 == 1) {
+      // 1x2 * 2x4 -> 1x4.
+      int8x8_t lhs = vld1_s8(lhs_ptr);
+      lhs_ptr += 8;
+      int8x8x2_t lhs_uzp = vuzp_s8(lhs, vzero);
+
+      acc[0] = vdotq_lane_s32(acc[0], rhs_0, lhs_uzp.val[0], 0);
+      acc[0] = vdotq_lane_s32(acc[0], rhs_1, lhs_uzp.val[1], 0);
+    }
+  }
+
+  for (int i = 0; i < M0; i++) {
+    acc[i] = vshrq_n_s32(acc[i], 4);
+    if (params->flags & IREE_UK_FLAG_MMT4D_ACCUMULATE) {
```

Comment:
Move this `if` branch outside of the enclosing `for`, duplicating the `for`. The compiler might perform that optimization on its own, or not; this move will help it make the right decision. That's because the optimal instruction sequence looks quite different between the two `if` cases here. Part of this is to use more registers and have some space between load and use, and part of this is that in the accumulating case, you want to use "right-shift-and-accumulate" instructions, SSRA.  The Arm intrinsics search seems to be broken as it doesn't return results for "right shift and accumulate", but if you search instead for SSRA, it works and returns these intrinsics:
https://developer.arm.com/architectures/instruction-sets/intrinsics/#q=SSRA
E.g. `vsraq_n_s32`.

Reply:
Thanks for the tip. Updated.